assembly/src/json.ts

```diff
@@ -559,6 +592,153 @@ export namespace JSON {
+            isKey = false;
+            break;
+          }
+        }
+      }
+    } else if (char === quoteCode) {
+      let escaping = false;
+      for (
+        let stringValueIndex = ++outerLoopIndex;
+        stringValueIndex < data.length - 1;
+        stringValueIndex++
+      ) {
+        const char = unsafeCharCodeAt(data, stringValueIndex);
+        if (char === backSlashCode && !escaping) {
+          escaping = true;
+        } else {
+          if (
+            char === quoteCode && !escaping
+          ) {
+            if (isKey === false) {
+              key.reinst(data, outerLoopIndex, stringValueIndex);
+              isKey = true;
+            } else {              
+              if (isString<valueof<T>>()) {
+                map.set(parseMapKey<indexof<T>>(key), data.slice(outerLoopIndex, stringValueIndex));
+              }
+              isKey = false;
+            }
+            outerLoopIndex = ++stringValueIndex;
+            break;
+          }
+          escaping = false;
+        }
+      }
+    } else if (
+      char == nCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === uCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === lCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === lCode) {
+      if (isNullable<valueof<T>>()) {
+        map.set(parseMapKey<indexof<T>>(key), null);
+      }
+      isKey = false;
+    } else if (
+      char === tCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === rCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === uCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === eCode
+    ) {
+      if (isBoolean<valueof<T>>()) {
+        map.set(parseMapKey<indexof<T>>(key), true);
+      }
+      isKey = false;
+    } else if (
+      char === fCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === aCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === lCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === sCode &&
+      unsafeCharCodeAt(data, ++outerLoopIndex) === eCode
+    ) {
+      if (isBoolean<valueof<T>>()) {
+        map.set(parseMapKey<indexof<T>>(key), false);
+      }
+      isKey = false;
+    } else if ((char >= 48 && char <= 57) || char === 45) {
+      let numberValueIndex = ++outerLoopIndex;
+      for (; numberValueIndex < data.length; numberValueIndex++) {
+        const char = unsafeCharCodeAt(data, numberValueIndex);
+        if (char === colonCode || char === commaCode || char === rightBraceCode || isSpace(char)) {
+          if (isFloat<valueof<T>>() || isInteger<valueof<T>>()) {
+            map.set(parseMapKey<indexof<T>>(key), parseNumber<valueof<T>>(data.slice(outerLoopIndex - 1, numberValueIndex)));
+          }          
+          outerLoopIndex = numberValueIndex;
+          isKey = false;
+          break;
+        }
+      }
+    }
+  }
+
+  return map;
+}
+
+//@ts-ignore: Decorator
+@inline function parseMapKey<T>(key: Virtual<string>): T {
+  const k = key.copyOut();
```

Comment:
@JairusSW - is `copyOut` the best way to handle this?  And does that defeat the purpose of it being virtual to begin with?  If there's a better way, please let me know.  Thanks.

Reply:
Using `copyOut()` here is going to be your best option. That is, unless I overhaul each distinct parser to accept `Virtual<T>` as input, which is not the case at the moment. `copyOut()` is appropriate here.

Reply:
That's fine. I can just use `copyOut` for now.  I think the perf is still reasonable.